File: commons/dto/src/main/java/net/blueshell/dto/MembershipDTO.java
package net.blueshell.dto;

import jakarta.validation.constraints.NotBlank;
import lombok.Data;
import lombok.EqualsAndHashCode;
import net.blueshell.common.enums.MemberType;

import java.util.Date;

@Data
@EqualsAndHashCode(callSuper = false)
public class MembershipDTO extends BaseDTO {

    private Long id;

    @NotBlank
    private Long userId;

    private MemberType memberType;

    private Date date;

    private String city;

    private String country;

    private FileDTO signature;

    private Date startDate;

    private Date endDate;
}

File: api/src/main/kotlin/net/blueshell/api/feature/membership/dto/MembershipDTO.kt
package net.blueshell.api.feature.membership.dto

import io.swagger.v3.oas.annotations.media.Schema
import jakarta.validation.constraints.NotNull
import jakarta.validation.constraints.PastOrPresent
import net.blueshell.api.shared.enums.MemberType
import net.blueshell.api.shared.dto.AuditedAutoIdDTO
import net.blueshell.api.shared.validation.date.Today
import net.blueshell.api.shared.validation.group.Administration
import net.blueshell.api.shared.validation.group.Creation
import net.blueshell.api.feature.membership.validation.NoExistingMembershipForUserId
import org.springframework.format.annotation.DateTimeFormat
import java.time.LocalDate

@Schema(name = "Membership")
data class MembershipDTO(
    @field:NotNull
    @field:NoExistingMembershipForUserId
    var userId: Long? = null,

    @field:NotNull(groups = [Administration::class])
    var memberType: MemberType? = null,

    @field:NotNull(groups = [Creation::class])
    var city: String? = null,

    @field:NotNull(groups = [Creation::class])
    var country: String? = null,

    @field:DateTimeFormat(iso = DateTimeFormat.ISO.DATE)
    @field:PastOrPresent(groups = [Administration::class])
    @field:Today(groups = [Creation::class])
    var startDate: LocalDate? = null,

    @field:DateTimeFormat(iso = DateTimeFormat.ISO.DATE)
    var endDate: LocalDate? = null,

    @field:NotNull
    var incasso: Boolean = false
) : AuditedAutoIdDTO()
